feat(file): add support for `import` content type

Vaneixus opened the pull request:
Signed-off-by: Marco Attia <54147992+Vaneixus@users.noreply.github.com>

### Contributor's Note
<!--- 
Please mark the following items with an [x] if they apply to your PR.
Leave the [ ] if they are not applicable, or if you have not completed the item.
--->
- [x] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [x] I have added / updated acceptance tests in `/fwprovider/tests` for any new or updated resources / data sources.
- [x] I have ran `make example` to verify that the change works as expected.

<!---
You can find more information about coding conventions and local testing in the [CONTRIBUTING.md](https://github.com/bpg/terraform-provider-proxmox/blob/main/CONTRIBUTING.md) file.

If you are unsure how to run `make example`, see [Deploying the example resources](https://github.com/bpg/terraform-provider-proxmox?tab=readme-ov-file#deploying-the-example-resources) section in README.
--->

<!--
*IF* your code contains breaking changes make sure to add `!` to the end of commit type, e.g.:
```
    feat(vm)!: add support for new feature 
```
Also, uncomment the section just below, and add a description of the breaking change. 
--->

<!---
#### ⚠ BREAKING CHANGES

>>> Put your description here <<<
--->

### Proof of Work
<!--- 
Please add screenshots, logs, or other relevant information that demonstrates the change works as expected.
--->

```hcl
resource "proxmox_virtual_environment_download_file" "latest_debian_12_bookworm_qcow2" {
  content_type = "import"
  datastore_id = "local"
  file_name    = "debian-12-generic-amd64.qcow2"
  node_name    = "pve"
  url          = "https://cloud.debian.org/images/cloud/bookworm/latest/debian-12-generic-amd64.qcow2"
}
```

<img width="561" alt="Screenshot 2025-06-27 at 9 28 27 PM" src="https://github.com/user-attachments/assets/f41a316a-ab08-4f32-bd43-092ff97601c6" />

<!--- Please keep this note for the community --->
### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request
<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates #1913 

<!--- Release note for [CHANGELOG](https://github.com/bpg/terraform-provider-proxmox/blob/main/CHANGELOG.md) will be created automatically using the PR's title, update it accordingly. --->

<!---
BEGIN_COMMIT_OVERRIDE
feat(file): add support for `import` content type (#1983)
END_COMMIT_OVERRIDE
--->

bpg commented on fwprovider/nodes/resource_download_file.go:
```diff
@@ -157,23 +157,22 @@ func (r *downloadFileResource) Schema(
 				},
 			},
 			"file_name": schema.StringAttribute{
-				Description: "The file name. If not provided, it is calculated " +
-					"using `url`. PVE will raise 'wrong file extension' error for some popular " +
-					"extensions file `.raw` or `.qcow2`. Workaround is to use e.g. `.img` instead.",
```
is there any particular reason for removing this note?

Vaneixus replied:
as for the first comment, it's mostly because 8.4 can import directly thus a workaround isn't necessary, but I see now that it might be better to keep it because 8.3 and prior versions do not support direct import, so I guess we can keep it.

bpg commented on proxmoxtf/resource/file.go:
```diff
@@ -638,22 +664,27 @@ func fileGetContentType(d *schema.ResourceData) (*string, diag.Diagnostics) {
 			strings.HasSuffix(sourceFilePath, ".tar.xz") {
 			contentType = "vztmpl"
+			// For Proxmox VE 8.4 and later, we can import VM images to  the "import" content type.
+		} else if releaseMajor >= 8 && releaseMinor > 4 && (strings.HasSuffix(sourceFilePath, ".qcow2") ||
```
should it be `releaseMinor >= 4` ?

bpg commented on proxmoxtf/resource/file.go:
```diff
@@ -638,22 +664,27 @@ func fileGetContentType(d *schema.ResourceData) (*string, diag.Diagnostics) {
 
 			switch ext {
-			case "img", "iso":
+			case "iso":
```
same here, not sure why `img` was removed. I think we should keep this code for backward compatibility.

bpg reviewed (commented):
Thanks for the feature @Vaneixus! ❤️ 

I have some comments about handling `img` extension, not sure why you want to remove this use case.

I've also took a liberty to refactor the version retrieval and parsing code to make it more reusable. There are few other places in the provider where we can add some capability checks (SDE is one that comes to my mind)

Vaneixus commented:
> Thanks for the feature @Vaneixus! ❤️
> 
> I have some comments about handling `img` extension, not sure why you want to remove this use case.
> 
> I've also took a liberty to refactor the version retrieval and parsing code to make it more reusable. There are few other places in the provider where we can add some capability checks (SDE is one that comes to my mind)

As for the version checks, yeah, I had a commit on that already. Just didn't push it yet because I didn't have time to test it on my PVE install as of yet.

Vaneixus commented:
I think I finished everything? I added new tests, fixed the documentation issues and ran the make example command. It currently hangs on some data sources(might just be the test VM, it's running on an unstable/slow ceph cluster), but for the import action done by the new examples, it works. I will have to take another look at it. other than that, it should be ready to merge.

Vaneixus commented:
I spent a few hours yesterday investigating the `make example`  command. it is failing at creating some of the VMs. I am not sure why. I do know however that one of the lxc images used for testing has been deleted from it's online repository. The ubuntu noble lxc image from 2024-07-25 to be exact. Later builds of the image do exist. However, the various exemples I had created did succeed at downloading the .qcow2 image as is into the import store. I haven't seen any tests for the upload tho, so I refrained from testing that, because I did not know if it is intentional or not. I shall include some logs and the fixes I had applied whenever I get back home. The apply also still hangs at the creation of other VMs(example trunk, example template and one another), even when they're already done being created on proxmox.

bpg commented:
> I spent a few hours yesterday investigating the `make example` command. it is failing at creating some of the VMs. I am not sure why.

Yeah, I noticed that bad behaviour too, [made some fixes](https://github.com/bpg/terraform-provider-proxmox/pull/2007) a few days back. 
No worries, i'll have time later today to go over the reviews, will run some more tests on your PR and merge it if everything works fine.

bpg commented:
> ... to include the requirements in 'README.md'. for example, there's no mention about enabling VLAN-aware for `vmbr0` which is a requirement for running the `make example` command.

[There is](https://github.com/bpg/terraform-provider-proxmox/blob/main/README.md#prerequisites) :)

bpg commented on docs/resources/virtual_environment_download_file.md:
```diff
@@ -27,6 +27,16 @@ resource "proxmox_virtual_environment_download_file" "release_20231228_debian_12
+resource "proxmox_virtual_environment_download_file" "release_20231228_debian_12_bookworm_qcow2" {
+  content_type       = "import"
+  datastore_id       = "local"
+  file_name          = "debian-12-generic-amd64-20231228-1609"
```
here and in other places, i think it should have .qcow extension

bpg reviewed (approved):
Thank you for the feature @Vaneixus!

LGTM! 🚀
